lib/date_time/date_time_core.c

```diff
@@ -53,26 +59,51 @@ static void date_time_core_notify_event(enum date_time_evt_type time_source)
 	}
 }
 
-static void date_time_core_schedule_update(bool check_pending)
+static int date_time_core_schedule_work(int interval)
 {
-	/* (Re-)schedule time update work
-	 * if periodic updates are requested in the configuration.
+	/* If a scheduled update is blocking reschedules, exit.
+	 * Otherwise set the reschedule_blocked flag to true, then proceed with the reschedule.
 	 */
-	if (CONFIG_DATE_TIME_UPDATE_INTERVAL_SECONDS > 0) {
-		/* Don't reschedule time update work in some cases,
-		 * e.g. if time is not obtained and the work is already pending.
-		 */
-		if (check_pending && k_work_delayable_is_pending(&date_time_update_work)) {
-			return;
-		}
 
+	if (atomic_set(&reschedule_blocked, true)) {
+		LOG_DBG("Requested date-time update not scheduled, another is already scheduled");
+		return -EAGAIN;
+	}
+
+	k_work_reschedule_for_queue(&date_time_work_q, &date_time_update_work, K_SECONDS(interval));
+
+	return 0;
+}
+
+static void date_time_core_schedule_update(void)
+{
+	if (CONFIG_DATE_TIME_UPDATE_INTERVAL_SECONDS <= 0) {
+		LOG_DBG("Skipping requested date time update, periodic requests are not enabled");
+		return;
+	}
+
+	/* Reset the retry counter since this will be a normal update. */
+	atomic_set(&retry_count, 0);
+
+	if (date_time_core_schedule_work(CONFIG_DATE_TIME_UPDATE_INTERVAL_SECONDS) == 0) {
 		LOG_DBG("New periodic date time update in: %d seconds",
 			CONFIG_DATE_TIME_UPDATE_INTERVAL_SECONDS);
+	}
+}
+
+static void date_time_core_schedule_retry(void)
+{
+	/* If this is the last allowable retry, or retries are not enabled
+	 * (CONFIG_DATE_TIME_RETRY_COUNT==0), switch to performing a normal update.
+	 */
+	if (atomic_inc(&retry_count) >= CONFIG_DATE_TIME_RETRY_COUNT) {
+		date_time_core_schedule_update();
+		return;
+	}
 
-		k_work_reschedule_for_queue(
-			&date_time_work_q,
-			&date_time_update_work,
-			K_SECONDS(CONFIG_DATE_TIME_UPDATE_INTERVAL_SECONDS));
+	if (date_time_core_schedule_work(CONFIG_DATE_TIME_RETRY_INTERVAL_SECONDS) == 0) {
```

Comment:
Did you test that this works nicely with `CONFIG_DATE_TIME_RETRY_INTERVAL_SECONDS=0` and `CONFIG_DATE_TIME_RETRY_COUNT` > 0? It should but worth double checking.

Reply:
I don't think I tested that specific combo, I will add DNM and test tomorrow

Reply:
Another task came up suddenly which superseded this one, so I did not wind up testing. I will test today.

In the mean time, I believe I have addressed your review concerns, so re-requesting review

Reply:
I don't have other comments. I was waiting for this test result.

Reply:
Thanks for the approval -- looks like it works as expected! (immediate retry) But I noticed a typo in my commit messages, so I will be force-pushing an update to that.